File: frontend/src/types/fullhand.ts
// 完整牌局模拟类型定义
// V1.1 新增

export type Street = 'preflop' | 'flop' | 'turn' | 'river' | 'showdown';
export type GameStatus = 'INIT' | 'PREFLOP' | 'FLOP_DECISION' | 'FAST_FORWARD' | 'ENDED';
export type PotType = 'SRP' | '3BP' | '4BP';
export type IPOOP = 'IP' | 'OOP';
export type SPRBucket = 'LOW' | 'MED' | 'HIGH';
export type HandBucket = 'made_strong' | 'made_medium' | 'made_weak' | 'draw_strong' | 'draw_weak' | 'air';
export type Grade = 'PERFECT' | 'ACCEPTABLE' | 'WRONG';

export interface PlayerState {
  seat: number;
  position: string; // UTG, MP, CO, BTN, SB, BB
  stack: number;
  in_hand: boolean;
  committed_this_street: number;
  total_committed: number;
  hole_cards?: string[];
  is_hero: boolean;
  is_active: boolean;
}

export interface ActionRecord {
  street: string;
  seat: number;
  position: string;
  action: string;
  amount?: number;
  pot_after: number;
  timestamp?: string;
}

export interface KeySpotInfo {
  street: string;
  context_id: string;
  pot_type?: PotType;
  ip_oop?: IPOOP;
  spr_bucket?: SPRBucket;
  board?: string[];
  hero_hand: string;
  hero_hand_bucket?: HandBucket;
  legal_actions: string[];
  strategy?: Record<string, number>;
  best_action: string;
}

export interface GameState {
  status: GameStatus;
  street: Street;
  players: PlayerState[];
  button_seat: number;
  sb_seat: number;
  bb_seat: number;
  pot: number;
  current_bet: number;
  community_cards: string[];
  to_act_seat?: number;
  hero_seat: number;
  hero_cards: string[];
  min_raise_to?: number;
}

export interface FullHandSession {
  hand_id: number;
  state: GameState;
  legal_actions: string[];
  action_log: ActionRecord[];
  is_key_spot: boolean;
  key_spot_info?: KeySpotInfo;
}

export interface FullHandStartRequest {
  table_type: string;
  stack_bb: number;
  ai_level: string;
  replay_seed?: string;
}

export interface FullHandActRequest {
  hand_id: number;
  action: string;
  amount?: number;
}

export interface FullHandActResponse {
  state: GameState;
  legal_actions: string[];
  is_key_spot: boolean;
  key_spot_info?: KeySpotInfo;
  final_result?: {
    result_bb: number;
    ended_by: string;
  };
  review_payload?: FullHandReview;
}

export interface PreflopSpotReview {
  context_id: string;
  strategy?: Record<string, number>;
  user_action: string;
  user_action_prob?: number;
  best_action: string;
  grade: Grade;
}

export interface FlopSpotReview {
  context_id: string;
  pot_type: PotType;
  ip_oop: IPOOP;
  spr_bucket: SPRBucket;
  board: string[];
  hero_hand: string;
  hero_hand_bucket: HandBucket;
  strategy?: Record<string, number>;
  legal_actions: string[];
  user_action: string;
  user_action_prob?: number;
  best_action: string;
  grade: Grade;
  explanation: string;
}

export interface FullHandReview {
  hand_id: number;
  result_bb: number;
  ended_by: string;
  action_log: ActionRecord[];
  preflop_spot?: PreflopSpotReview;
  flop_spot?: FlopSpotReview;
  can_replay: boolean;
}

export interface FullHandStats {
  total_hands: number;
  total_result_bb: number;
  avg_result_bb: number;
  preflop_accuracy?: number;
  flop_accuracy?: number;
  today_hands: number;
  today_remaining: number;
  is_pro: boolean;
}

// 下注尺度
export interface BetSizes {
  bet33: number;
  bet75: number;
  bet125: number;
}

// 动作标签映射
export const ACTION_LABELS: Record<string, string> = {
  fold: '弃牌',
  check: '过牌',
  call: '跟注',
  bet: '下注',
  raise: '加注',
  allin: '全下',
  bet33: '33% 底池',
  bet75: '75% 底池',
  bet125: '125% 底池',
  raise75: '75% 底池',
  raise125: '125% 底池',
  sb: '小盲',
  bb: '大盲',
};

// 位置颜色
export const POSITION_COLORS: Record<string, string> = {
  UTG: 'text-red-400',
  MP: 'text-orange-400',
  CO: 'text-yellow-400',
  BTN: 'text-green-400',
  SB: 'text-blue-400',
  BB: 'text-purple-400',
};

// 评级颜色
export const GRADE_COLORS: Record<Grade, string> = {
  PERFECT: 'text-green-400',
  ACCEPTABLE: 'text-yellow-400',
  WRONG: 'text-red-400',
};

// 评级图标
export const GRADE_ICONS: Record<Grade, string> = {
  PERFECT: '✅',
  ACCEPTABLE: '⚠️',
  WRONG: '❌',
};

File: backend/app/schemas/fullhand.py
"""
完整牌局模拟的 Pydantic Schemas
V1.1 新增
"""
from typing import List, Optional, Dict, Any
from pydantic import BaseModel
from datetime import datetime


# ========== 基础数据结构 ==========

class PlayerState(BaseModel):
    """玩家状态"""
    seat: int
    position: str  # UTG, MP, CO, BTN, SB, BB
    stack: float
    in_hand: bool
    committed_this_street: float
    total_committed: float
    hole_cards: Optional[List[str]] = None  # 仅 Hero 可见
    is_hero: bool
    is_active: bool  # 本轮是否还能行动


class ActionRecord(BaseModel):
    """行动记录"""
    street: str
    seat: int
    position: str
    action: str  # fold, check, call, bet, raise, allin, sb, bb, deal
    amount: Optional[float] = None
    pot_after: float
    timestamp: Optional[str] = None


class KeySpotInfo(BaseModel):
    """关键点信息"""
    street: str
    context_id: str
    pot_type: Optional[str] = None  # SRP, 3BP, 4BP
    ip_oop: Optional[str] = None  # IP, OOP
    spr_bucket: Optional[str] = None  # LOW, MED, HIGH
    board: Optional[List[str]] = None
    hero_hand: str
    hero_hand_bucket: Optional[str] = None
    legal_actions: List[str]
    strategy: Optional[Dict[str, float]] = None  # Pro 才展示完整
    best_action: str


# ========== 请求 Schemas ==========

class FullHandStartRequest(BaseModel):
    """开始一局请求"""
    table_type: str = "6max"
    stack_bb: int = 100
    ai_level: str = "standard"
    replay_seed: Optional[str] = None  # Pro 重打时使用


class FullHandActRequest(BaseModel):
    """执行动作请求"""
    hand_id: int
    action: str  # fold, check, call, bet, raise, allin
    amount: Optional[float] = None  # bet/raise 时需要


class FullHandReplayRequest(BaseModel):
    """重打请求"""
    hand_id: int


# ========== 响应 Schemas ==========

class GameState(BaseModel):
    """游戏状态"""
    status: str
    street: str
    players: List[PlayerState]
    button_seat: int
    sb_seat: int
    bb_seat: int
    pot: float
    current_bet: float
    community_cards: List[str]
    to_act_seat: Optional[int] = None
    hero_seat: int
    hero_cards: List[str]
    min_raise_to: Optional[float] = None


class FullHandStartResponse(BaseModel):
    """开始一局响应"""
    hand_id: int
    state: GameState
    legal_actions: List[str]
    action_log: List[ActionRecord]
    is_key_spot: bool = False
    key_spot_info: Optional[KeySpotInfo] = None


class FullHandActResponse(BaseModel):
    """执行动作响应"""
    state: GameState
    legal_actions: List[str]
    is_key_spot: bool
    key_spot_info: Optional[KeySpotInfo] = None
    final_result: Optional[Dict[str, Any]] = None  # 如果游戏结束
    review_payload: Optional[Dict[str, Any]] = None  # 如果游戏结束


class PreflopSpotReview(BaseModel):
    """翻前关键点复盘"""
    context_id: str
    strategy: Optional[Dict[str, float]] = None  # Free 裁剪
    user_action: str
    user_action_prob: Optional[float] = None  # Free 裁剪
    best_action: str
    grade: str  # PERFECT, ACCEPTABLE, WRONG


class FlopSpotReview(BaseModel):
    """翻牌关键点复盘"""
    context_id: str
    pot_type: str
    ip_oop: str
    spr_bucket: str
    board: List[str]
    hero_hand: str
    hero_hand_bucket: str
    strategy: Optional[Dict[str, float]] = None  # Free 裁剪
    legal_actions: List[str]
    user_action: str
    user_action_prob: Optional[float] = None  # Free 裁剪
    best_action: str
    grade: str
    explanation: str


class PlayerShowdownResult(BaseModel):
    """玩家摊牌结果"""
    seat: int
    position: str
    hole_cards: List[str]
    is_hero: bool
    in_hand: bool
    total_committed: float
    score: int
    hand_name: str
    is_winner: bool


class ShowdownAnalysis(BaseModel):
    """摊牌详细分析"""
    community_cards: List[str]
    pot: float
    players: List[PlayerShowdownResult]
    winner_analysis: Optional[Dict[str, Any]] = None
    explanation: str


class FullHandReviewResponse(BaseModel):
    """复盘响应"""
    hand_id: int
    result_bb: float
    ended_by: str
    action_log: List[ActionRecord]
    preflop_spot: Optional[PreflopSpotReview] = None
    flop_spot: Optional[FlopSpotReview] = None
    can_replay: bool  # 是否支持重打 (Pro)
    showdown_analysis: Optional[ShowdownAnalysis] = None  # 摊牌详细分析


class FullHandHistoryItem(BaseModel):
    """历史记录项"""
    hand_id: int
    stack_bb: int
    result_bb: float
    ended_by: str
    has_flop_spot: bool
    flop_grade: Optional[str] = None
    created_at: datetime


class FullHandStatsResponse(BaseModel):
    """统计响应"""
    total_hands: int
    total_result_bb: float
    avg_result_bb: float
    preflop_accuracy: Optional[float] = None
    flop_accuracy: Optional[float] = None
    today_hands: int
    today_remaining: int  # Free 用户今日剩余局数
    is_pro: bool
